// zmachine-core/dist/utils/constants.js
// general purpose constants
// constants for testing/setting bits
export const BIT_0 = 1;
export const BIT_1 = 2;
export const BIT_2 = 4;
export const BIT_3 = 8;
export const BIT_4 = 16;
export const BIT_5 = 32;
export const BIT_6 = 64;
export const BIT_7 = 128;
export const BIT_8 = 256;
export const BIT_9 = 512;
export const BIT_10 = 1024;
export const BIT_11 = 2048;
export const BIT_12 = 4096;
export const BIT_13 = 8192;
export const BIT_14 = 16384;
export const BIT_15 = 32768;
export const BIT_16 = 65536;
export const BIT_17 = 131072;
export const BIT_18 = 262144;
export const BIT_19 = 524288;
export const BIT_20 = 1048576;
export const BIT_21 = 2097152;
export const BIT_22 = 4194304;
export const BIT_23 = 8388608;
export const BIT_24 = 16777216;
export const BIT_25 = 33554432;
export const BIT_26 = 67108864;
export const BIT_27 = 134217728;
export const BIT_28 = 268435456;
export const BIT_29 = 536870912;
export const BIT_30 = 1073741824;
export const BIT_31 = 2147483648;
export const BIT_32 = 4294967296;
export const BIT_33 = 8589934592;
export const BIT_34 = 17179869184;
export const BIT_35 = 34359738368;
export const BIT_36 = 68719476736;
export const BIT_37 = 137438953472;
export const BIT_38 = 274877906944;
export const BIT_39 = 549755813888;
export const BIT_40 = 1099511627776;
export const BIT_41 = 2199023255552;
export const BIT_42 = 4398046511104;
export const BIT_43 = 8796093022208;
export const BIT_44 = 17592186044416;
export const BIT_45 = 35184372088832;
export const BIT_46 = 70368744177664;
export const BIT_47 = 140737488355328;
export const BIT_48 = 281474976710656;
export const BITS = [
    BIT_0, BIT_1, BIT_2, BIT_3, BIT_4, BIT_5, BIT_6, BIT_7, BIT_8, BIT_9,
    BIT_10, BIT_11, BIT_12, BIT_13, BIT_14, BIT_15, BIT_16, BIT_17, BIT_18, BIT_19,
    BIT_20, BIT_21, BIT_22, BIT_23, BIT_24, BIT_25, BIT_26, BIT_27, BIT_28, BIT_29,
    BIT_30, BIT_31, BIT_32, BIT_33, BIT_34, BIT_35, BIT_36, BIT_37, BIT_38, BIT_39,
    BIT_40, BIT_41, BIT_42, BIT_43, BIT_44, BIT_45, BIT_46, BIT_47, BIT_48,
];
// constants for masking bytes
export const BYTE_MASK_0 = 255;
export const BYTE_MASK_1 = 65280;
export const BYTE_MASK_2 = 16711680;
export const BYTE_MASK_3 = 4278190080;
export const BYTE_MASK_4 = 1095216660480;
export const BYTE_MASK_5 = 280375465082880;
export const WORD_MASK = 65535;
