// zmachine-core/dist/resources/object.js
import Memory from '../state/memory';
import Header from './header';
import { BIT_8, BIT_16, BYTE_MASK_0 } from '../utils/constants';
export default class Objects {
    static getChild(object) {
        const version = Header.version();
        if (version < 4) {
            return this.getChildV1(object);
        }
        return this.getChildV4(object);
    }
    static getChildV1(object) {
        if (object === 0) {
            return 0;
        }
        const objectAddr = this.getObjectAddrV1(object);
        return Memory.getByte(objectAddr + 6);
    }
    static getChildV4(object) {
        if (object === 0) {
            return 0;
        }
        const objectAddr = this.getObjectAddrV4(object);
        return Memory.getWord(objectAddr + 10);
    }
    static setChildV1(object, child) {
        if (object === 0) {
            throw new Error('setChildV1 on object zero');
        }
        const objectAddr = this.getObjectAddrV1(object);
        Memory.setByte(objectAddr + 6, child);
    }
    static setChildV4(object, child) {
        if (object === 0) {
            throw new Error('setChildV4 on object zero');
        }
        const objectAddr = this.getObjectAddrV4(object);
        Memory.setWord(objectAddr + 10, child);
    }
    static getDefaultPropertyV1(property) {
        if (property < 1 || property > 31) {
            throw new Error(`property out of range ${property}`);
        }
        const objectTable = Header.objectTable();
        return Memory.getWord(objectTable + (2 * (property - 1)));
    }
    static getDefaultPropertyV4(property) {
        if (property < 1 || property > 63) {
            throw new Error(`property out of range ${property}`);
        }
        const objectTable = Header.objectTable();
        return Memory.getWord(objectTable + (2 * (property - 1)));
    }
    static getNameWords(object) {
        if (object === 0) {
            throw new Error('getNameWords on object zero');
        }
        let propertyTableAddr;
        const version = Header.version();
        if (version < 4) {
            const objectAddr = this.getObjectAddrV1(object);
            propertyTableAddr = Memory.getWord(objectAddr + 7);
        }
        else {
            const objectAddr = this.getObjectAddrV4(object);
            propertyTableAddr = Memory.getWord(objectAddr + 12);
        }
        const nameSize = Memory.getByte(propertyTableAddr);
        const nameWords = [];
        for (let i = 0; i < nameSize; i += 1) {
            nameWords.push(Memory.getWord(propertyTableAddr + 1 + (2 * i)));
        }
        return nameWords;
    }
    // get the start byte of entry for an object, version 1-3
    static getObjectAddrV1(object) {
        if (object < 1 || object >= BIT_8) {
            throw new Error(`object is out of range ${object}`);
        }
        const objectTable = Header.objectTable();
        // 62 bytes of defaults, plus 9 for each object (except object 0)
        return objectTable + 53 + (9 * object);
    }
    // get the start byte of entry for an object, versions 4+
    static getObjectAddrV4(object) {
        if (object < 1 || object >= BIT_16) {
            throw new Error(`object is out of range ${object}`);
        }
        const objectTable = Header.objectTable();
        // 126 bytes of defaults, plus 14 for each object (except object 0)
        return objectTable + 112 + (14 * object);
    }
    static getParent(object) {
        const version = Header.version();
        if (version < 4) {
            return this.getParentV1(object);
        }
        return this.getParentV4(object);
    }
    static getParentV1(object) {
        if (object === 0) {
            return 0;
        }
        const objectAddr = this.getObjectAddrV1(object);
        return Memory.getByte(objectAddr + 4);
    }
    static getParentV4(object) {
        if (object === 0) {
            return 0;
        }
        const objectAddr = this.getObjectAddrV4(object);
        return Memory.getWord(objectAddr + 6);
    }
    static setParentV1(object, parent) {
        if (object === 0) {
            throw new Error('setParentV1 on object zero');
        }
        const objectAddr = this.getObjectAddrV1(object);
        Memory.setByte(objectAddr + 4, parent);
    }
    static setParentV4(object, parent) {
        if (object === 0) {
            throw new Error('setParentV4 on object zero');
        }
        const objectAddr = this.getObjectAddrV4(object);
        Memory.setWord(objectAddr + 6, parent);
    }
    static getProp(object, property) {
        const version = Header.version();
        if (version < 4) {
            const propertyAddr = this.getPropertyAddrV1(object, property);
            if (propertyAddr === 0) {
                return this.getDefaultPropertyV1(property);
            }
            return this.getPropertyV1(propertyAddr);
        }
        const propertyAddr = this.getPropertyAddrV4(object, property);
        if (propertyAddr === 0) {
            return this.getDefaultPropertyV4(property);
        }
        return this.getPropertyV4(propertyAddr);
    }
    static getPropertyV1(propertyAddr) {
        const sizeByte = Memory.getByte(propertyAddr - 1);
        // eslint-disable-next-line no-bitwise
        const size = ((sizeByte & 224) >> 5) + 1;
        if (size === 1) {
            return Memory.getByte(propertyAddr);
        }
        if (size === 2) {
            return Memory.getWord(propertyAddr);
        }
        throw new Error(`Unexpected size ${size}`);
    }
    static getPropertyV4(propertyAddr) {
        const sizeByte = Memory.getByte(propertyAddr - 1);
        let size;
        // eslint-disable-next-line no-bitwise
        if ((sizeByte & 128) === 0) {
            // eslint-disable-next-line no-bitwise
            if ((sizeByte & 64) === 0) {
                size = 1;
            }
            else {
                size = 2;
            }
        }
        else {
            // eslint-disable-next-line no-bitwise
            size = sizeByte & 63;
            if (size === 0) {
                size = 64;
            }
        }
        if (size === 1) {
            return Memory.getByte(propertyAddr);
        }
        if (size === 2) {
            return Memory.getWord(propertyAddr);
        }
        throw new Error(`Unexpected size ${size}`);
    }
    static getNextProperty(object, property) {
        const version = Header.version();
        if (version < 4) {
            return this.getNextPropertyV1(object, property);
        }
        return this.getNextPropertyV4(object, property);
    }
    static getNextPropertyV1(object, property) {
        if (property < 0 || property > 31) {
            throw new Error(`property out of range ${property}`);
        }
        if (object === 0) {
            return 0;
        }
        const objectAddr = this.getObjectAddrV1(object);
        let propertyAddr = Memory.getWord(objectAddr + 7);
        const nameSize = Memory.getByte(propertyAddr);
        propertyAddr += 1 + (nameSize * 2);
        let sizeByte = Memory.getByte(propertyAddr);
        propertyAddr += 1;
        if (property === 0) {
            // eslint-disable-next-line no-bitwise
            return sizeByte & 31;
        }
        while (sizeByte !== 0) {
            // eslint-disable-next-line no-bitwise
            const num = sizeByte & 31;
            // eslint-disable-next-line no-bitwise
            const size = ((sizeByte & 224) >> 5) + 1;
            propertyAddr += size;
            sizeByte = Memory.getByte(propertyAddr);
            propertyAddr += 1;
            if (num === property) {
                // eslint-disable-next-line no-bitwise
                return sizeByte & 31;
            }
            if (num < property) {
                throw new Error('property not found');
            }
        }
        throw new Error('property not found');
    }
    static getNextPropertyV4(object, property) {
        if (property < 0 || property > 63) {
            throw new Error(`property out of range ${property}`);
        }
        if (object === 0) {
            return 0;
        }
        const objectAddr = this.getObjectAddrV4(object);
        let propertyAddr = Memory.getWord(objectAddr + 12);
        const nameSize = Memory.getByte(propertyAddr);
        propertyAddr += 1 + (nameSize * 2);
        let sizeByte1 = Memory.getByte(propertyAddr);
        propertyAddr += 1;
        if (property === 0) {
            // eslint-disable-next-line no-bitwise
            return sizeByte1 & 63;
        }
        while (sizeByte1 !== 0) {
            // eslint-disable-next-line no-bitwise
            const num = sizeByte1 & 63;
            let size;
            // eslint-disable-next-line no-bitwise
            if ((sizeByte1 & 128) === 0) {
                // eslint-disable-next-line no-bitwise
                if ((sizeByte1 & 64) === 0) {
                    size = 1;
                }
                else {
                    size = 2;
                }
            }
            else {
                const sizeByte2 = Memory.getByte(propertyAddr);
                propertyAddr += 1;
                // eslint-disable-next-line no-bitwise
                size = sizeByte2 & 63;
                if (size === 0) {
                    size = 64;
                }
            }
            propertyAddr += size;
            sizeByte1 = Memory.getByte(propertyAddr);
            propertyAddr += 1;
            if (num === property) {
                // eslint-disable-next-line no-bitwise
                return sizeByte1 & 63;
            }
            if (num < property) {
                throw new Error('property not found');
            }
        }
        throw new Error('property not found');
    }
    static getPropertyAddr(object, property) {
        const version = Header.version();
        if (version < 4) {
            return this.getPropertyAddrV1(object, property);
        }
        return this.getPropertyAddrV4(object, property);
    }
    static getPropertyAddrV1(object, property) {
        if (property < 1 || property > 31) {
            throw new Error(`property out of range ${property}`);
        }
        if (object === 0) {
            return 0;
        }
        const objectAddr = this.getObjectAddrV1(object);
        let propertyAddr = Memory.getWord(objectAddr + 7);
        const nameSize = Memory.getByte(propertyAddr);
        propertyAddr += 1 + (nameSize * 2);
        let sizeByte = Memory.getByte(propertyAddr);
        propertyAddr += 1;
        while (sizeByte !== 0) {
            // eslint-disable-next-line no-bitwise
            const num = sizeByte & 31;
            // eslint-disable-next-line no-bitwise
            const size = ((sizeByte & 224) >> 5) + 1;
            if (num === property) {
                return propertyAddr;
            }
            if (num < property) {
                return 0;
            }
            propertyAddr += size;
            sizeByte = Memory.getByte(propertyAddr);
            propertyAddr += 1;
        }
        return 0;
    }
    static getPropertyAddrV4(object, property) {
        if (property < 1 || property > 63) {
            throw new Error(`property out of range ${property}`);
        }
        if (object === 0) {
            return 0;
        }
        const objectAddr = this.getObjectAddrV4(object);
        let propertyAddr = Memory.getWord(objectAddr + 12);
        const nameSize = Memory.getByte(propertyAddr);
        propertyAddr += 1 + (nameSize * 2);
        let sizeByte1 = Memory.getByte(propertyAddr);
        propertyAddr += 1;
        while (sizeByte1 !== 0) {
            // eslint-disable-next-line no-bitwise
            const num = sizeByte1 & 63;
            let size;
            // eslint-disable-next-line no-bitwise
            if ((sizeByte1 & 128) === 0) {
                // eslint-disable-next-line no-bitwise
                if ((sizeByte1 & 64) === 0) {
                    size = 1;
                }
                else {
                    size = 2;
                }
            }
            else {
                const sizeByte2 = Memory.getByte(propertyAddr);
                propertyAddr += 1;
                // eslint-disable-next-line no-bitwise
                size = sizeByte2 & 63;
                if (size === 0) {
                    size = 64;
                }
            }
            if (num === property) {
                return propertyAddr;
            }
            if (num < property) {
                return 0;
            }
            propertyAddr += size;
            sizeByte1 = Memory.getByte(propertyAddr);
            propertyAddr += 1;
        }
        return 0;
    }
    static getPropertyLength(propertyAddress) {
        const version = Header.version();
        if (version < 4) {
            return this.getPropertyLengthV1(propertyAddress);
        }
        return this.getPropertyLengthV4(propertyAddress);
    }
    static getPropertyLengthV1(propertyAddress) {
        const sizeByte = Memory.getByte(propertyAddress - 1);
        // eslint-disable-next-line no-bitwise
        const size = ((sizeByte & 224) >> 5) + 1;
        return size;
    }
    static getPropertyLengthV4(propertyAddress) {
        const sizeByte = Memory.getByte(propertyAddress - 1);
        let size;
        // eslint-disable-next-line no-bitwise
        if ((sizeByte & 128) === 0) {
            // eslint-disable-next-line no-bitwise
            if ((sizeByte & 64) === 0) {
                size = 1;
            }
            else {
                size = 2;
            }
        }
        else {
            // eslint-disable-next-line no-bitwise
            size = sizeByte & 63;
            if (size === 0) {
                size = 64;
            }
        }
        return size;
    }
    static getSibling(object) {
        const version = Header.version();
        if (version < 4) {
            return this.getSiblingV1(object);
        }
        return this.getSiblingV4(object);
    }
    static getSiblingV1(object) {
        if (object === 0) {
            return 0;
        }
        const objectAddr = this.getObjectAddrV1(object);
        return Memory.getByte(objectAddr + 5);
    }
    static getSiblingV4(object) {
        if (object === 0) {
            return 0;
        }
        const objectAddr = this.getObjectAddrV4(object);
        return Memory.getWord(objectAddr + 8);
    }
    static setSiblingV1(object, sibling) {
        if (object === 0) {
            throw new Error('setSiblingV1 on object zero');
        }
        const objectAddr = this.getObjectAddrV1(object);
        Memory.setByte(objectAddr + 5, sibling);
    }
    static setSiblingV4(object, sibling) {
        if (object === 0) {
            throw new Error('setSiblingV4 on object zero');
        }
        const objectAddr = this.getObjectAddrV4(object);
        Memory.setWord(objectAddr + 8, sibling);
    }
    static putProp(object, property, value) {
        const version = Header.version();
        if (version < 4) {
            const propertyAddr = this.getPropertyAddrV1(object, property);
            if (propertyAddr === 0) {
                throw new Error('put_prop on non-existent property');
            }
            this.putPropertyV1(propertyAddr, value);
        }
        else {
            const propertyAddr = this.getPropertyAddrV4(object, property);
            if (propertyAddr === 0) {
                throw new Error('put_prop on non-existent property');
            }
            this.putPropertyV4(propertyAddr, value);
        }
    }
    static insertObject(object, destination) {
        const version = Header.version();
        if (version < 4) {
            this.insertObjectV1(object, destination);
        }
        else {
            this.insertObjectV4(object, destination);
        }
    }
    static insertObjectV1(object, destination) {
        if (object === 0) {
            return;
        }
        if (object < 1 || object >= BIT_8) {
            throw new Error('invalid object');
        }
        this.removeObjectV1(object);
        if (destination === 0) {
            return;
        }
        if (destination < 1 || destination >= BIT_8) {
            throw new Error('invalid destination');
        }
        const child = this.getChildV1(destination);
        this.setSiblingV1(object, child);
        this.setChildV1(destination, object);
        this.setParentV1(object, destination);
    }
    static insertObjectV4(object, destination) {
        if (object === 0) {
            return;
        }
        if (object < 1 || object >= BIT_16) {
            throw new Error('invalid object');
        }
        this.removeObjectV4(object);
        if (destination === 0) {
            return;
        }
        if (destination < 1 || destination >= BIT_16) {
            throw new Error('invalid destination');
        }
        const child = this.getChildV4(destination);
        this.setSiblingV4(object, child);
        this.setChildV4(destination, object);
        this.setParentV4(object, destination);
    }
    static removeObject(object) {
        const version = Header.version();
        if (version < 4) {
            this.removeObjectV1(object);
        }
        else {
            this.removeObjectV4(object);
        }
    }
    static removeObjectV1(object) {
        if (object === 0) {
            return;
        }
        if (object < 1 || object >= BIT_8) {
            throw new Error('invalid object');
        }
        const parent = this.getParentV1(object);
        if (parent === 0) {
            return;
        }
        const child = this.getChildV1(parent);
        if (object === child) {
            // assign sibling as child
            const sibling = this.getSiblingV1(object);
            this.setChildV1(parent, sibling);
        }
        else {
            // find object's older sibling
            let older = child;
            let next = this.getSiblingV1(older);
            while (next !== object) {
                older = next;
                next = this.getSiblingV1(older);
            }
            // assign younger sibling to older
            const younger = this.getSiblingV1(object);
            this.setSiblingV1(older, younger);
        }
        this.setParentV1(object, 0);
        this.setSiblingV1(object, 0);
    }
    static removeObjectV4(object) {
        if (object === 0) {
            return;
        }
        if (object < 1 || object >= BIT_16) {
            throw new Error('invalid object');
        }
        const parent = this.getParentV4(object);
        if (parent === 0) {
            return;
        }
        const child = this.getChildV4(parent);
        if (object === child) {
            // assign sibling as child
            const sibling = this.getSiblingV4(object);
            this.setChildV4(parent, sibling);
        }
        else {
            // find object's older sibling
            let older = child;
            let next = this.getSiblingV4(older);
            while (next !== object) {
                older = next;
                next = this.getSiblingV4(older);
            }
            // assign younger sibling to older
            const younger = this.getSiblingV4(object);
            this.setSiblingV4(older, younger);
        }
        this.setParentV4(object, 0);
        this.setSiblingV4(object, 0);
    }
    static putPropertyV1(propertyAddr, value) {
        if (value < 0 || value >= BIT_16) {
            throw new Error(`value is out of range ${value}`);
        }
        const size = this.getPropertyLengthV1(propertyAddr);
        if (size === 1) {
            // eslint-disable-next-line no-bitwise
            Memory.setByte(propertyAddr, value & BYTE_MASK_0);
        }
        else if (size === 2) {
            // eslint-disable-next-line no-bitwise
            Memory.setWord(propertyAddr, value);
        }
        else {
            throw new Error(`Unexpected size ${size}`);
        }
    }
    static putPropertyV4(propertyAddr, value) {
        if (value < 0 || value >= BIT_16) {
            throw new Error(`value is out of range ${value}`);
        }
        const size = this.getPropertyLengthV4(propertyAddr);
        if (size === 1) {
            // eslint-disable-next-line no-bitwise
            Memory.setByte(propertyAddr, value & BYTE_MASK_0);
        }
        else if (size === 2) {
            // eslint-disable-next-line no-bitwise
            Memory.setWord(propertyAddr, value);
        }
        else {
            throw new Error(`Unexpected size ${size}`);
        }
    }
    static clearAttr(object, attribute) {
        if (object === 0) {
            return;
        }
        const version = Header.version();
        let objectAddr;
        if (version < 4) {
            if (object < 1 || object >= BIT_8) {
                throw new Error('invalid object');
            }
            if (attribute < 0 || attribute >= 32) {
                throw new Error('invalid attribute');
            }
            objectAddr = this.getObjectAddrV1(object);
        }
        else {
            if (object < 1 || object >= BIT_16) {
                throw new Error('invalid object');
            }
            if (attribute < 0 || attribute >= 48) {
                throw new Error('invalid attribute');
            }
            objectAddr = this.getObjectAddrV4(object);
        }
        // eslint-disable-next-line no-bitwise
        const bitMask = 128 >> (attribute & 7);
        // eslint-disable-next-line no-bitwise
        const byteNum = (attribute & 56) >> 3;
        const byte = Memory.getByte(objectAddr + byteNum);
        // eslint-disable-next-line no-bitwise
        Memory.setByte(objectAddr + byteNum, byte & ~bitMask);
    }
    static setAttr(object, attribute) {
        if (object === 0) {
            return;
        }
        const version = Header.version();
        let objectAddr;
        if (version < 4) {
            if (object < 1 || object >= BIT_8) {
                throw new Error('invalid object');
            }
            if (attribute < 0 || attribute >= 32) {
                throw new Error('invalid attribute');
            }
            objectAddr = this.getObjectAddrV1(object);
        }
        else {
            if (object < 1 || object >= BIT_16) {
                throw new Error('invalid object');
            }
            if (attribute < 0 || attribute >= 48) {
                throw new Error('invalid attribute');
            }
            objectAddr = this.getObjectAddrV4(object);
        }
        // eslint-disable-next-line no-bitwise
        const bitMask = 128 >> (attribute & 7);
        // eslint-disable-next-line no-bitwise
        const byteNum = (attribute & 56) >> 3;
        const byte = Memory.getByte(objectAddr + byteNum);
        // eslint-disable-next-line no-bitwise
        Memory.setByte(objectAddr + byteNum, byte | bitMask);
    }
    static testAttr(object, attribute) {
        if (object === 0) {
            return 0;
        }
        const version = Header.version();
        let objectAddr;
        if (version < 4) {
            if (object < 1 || object >= BIT_8) {
                throw new Error('invalid object');
            }
            if (attribute < 0 || attribute >= 32) {
                throw new Error('invalid attribute');
            }
            objectAddr = this.getObjectAddrV1(object);
        }
        else {
            if (object < 1 || object >= BIT_16) {
                throw new Error('invalid object');
            }
            if (attribute < 0 || attribute >= 48) {
                throw new Error('invalid attribute');
            }
            objectAddr = this.getObjectAddrV4(object);
        }
        // eslint-disable-next-line no-bitwise
        const bitMask = 128 >> (attribute & 7);
        // eslint-disable-next-line no-bitwise
        const byteNum = (attribute & 56) >> 3;
        const byte = Memory.getByte(objectAddr + byteNum);
        // eslint-disable-next-line no-bitwise
        if ((byte & bitMask) === 0) {
            return 0;
        }
        return 1;
    }
}
