// zmachine-core/dist/resources/zscii.js
// module for translating from zscii to strings and back
import Memory from '../state/memory';
import Header from './header';
const defaultTranslationTable = [
    228, 246, 252, 196, 214, 220, 223, 187, 171, 235,
    239, 255, 203, 207, 225, 233, 237, 243, 250, 253,
    193, 201, 205, 211, 218, 221, 224, 232, 236, 242,
    249, 192, 200, 204, 210, 217, 226, 234, 238, 244,
    251, 194, 202, 206, 212, 219, 229, 197, 248, 216,
    227, 241, 245, 195, 209, 213, 230, 198, 231, 199,
    254, 240, 222, 208, 163, 339, 338, 161, 191,
];
const inputCharMap = {
    Enter: 13,
    Backspace: 8,
    Escape: 27,
    ArrowUp: 129,
    ArrowDown: 130,
    ArrowLeft: 131,
    ArrowRight: 132,
    F1: 133,
    F2: 134,
    F3: 135,
    F4: 136,
    F5: 137,
    F6: 138,
    F7: 139,
    F8: 140,
    F9: 141,
    F10: 142,
    F11: 143,
    F12: 144,
    Keypad0: 145,
    Keypad1: 146,
    Keypad2: 147,
    Keypad3: 148,
    Keypad4: 149,
    Keypad5: 150,
    Keypad6: 151,
    Keypad7: 152,
    Keypad8: 153,
    Keypad9: 154,
};
export default class ZSCII {
    // translates a number of zscii output codes to a string
    static outputZSCIIToString(zscii) {
        const charCodes = [];
        for (let i = 0; i < zscii.length; i += 1) {
            const code = zscii[i];
            if (code >= 32 && code < 127) {
                charCodes.push(code);
            }
            else if (code === 13) {
                // newline
                charCodes.push(10);
            }
            else if (code === 9) {
                // tab, replace with 5 spaces
                charCodes.push(32, 32, 32, 32, 32);
            }
            else if (code === 11) {
                // sentence space AKA space in the year 2020
                charCodes.push(32);
            }
            else if (code === 0) {
                // no op
            }
            else if (code >= 155 && code < 252) {
                // extra characters
                charCodes.push(this.getExtraCharacter(code));
            }
            else {
                throw new Error(`Unexpected output ZSCCI code ${code}`);
            }
        }
        return String.fromCharCode(...charCodes);
    }
    static getExtraCharacter(code) {
        if (code < 155 || code >= 252) {
            throw new Error('extra character out of range');
        }
        const table = Header.getHeaderExtensionAddr(3);
        if (table > 0) {
            const tableLength = Memory.getByte(table);
            if (code - 155 < tableLength) {
                return Memory.getWord(table + 1 + ((code - 155) * 2));
            }
            throw new Error(`code not defined in extra characters table ${code}`);
        }
        // use default translation table
        return defaultTranslationTable[code - 155];
    }
    static lookupExtraCharacter(charCode) {
        // returns 0 if this unicode character is not one the extra characters
        const table = Header.getHeaderExtensionAddr(3);
        if (table > 0) {
            const tableLength = Memory.getByte(table);
            for (let i = 0; i < tableLength; i += 1) {
                if (Memory.getWord(table + 1 + (i * 2)) === charCode) {
                    return 155 + i;
                }
            }
            return 0;
        }
        // use default translation table
        for (let i = 0; i < defaultTranslationTable.length; i += 1) {
            if (defaultTranslationTable[i] === charCode) {
                return 155 + i;
            }
        }
        return 0;
    }
    static stringToInputZSCII(keyPress) {
        // returns 0 on unrecognized key presses
        if (keyPress.length === 1) {
            const code = keyPress.charCodeAt(0);
            if (code >= 32 && code < 127) {
                // standard ASCII
                return code;
            }
            // check if this is one of this game's extra characters
            return this.lookupExtraCharacter(code);
        }
        return inputCharMap[keyPress] || 0;
    }
}
