// zmachine-core/dist/state/input.js
export const KEYBOARD_STREAM = 0;
export const FILE_STREAM = 1;
class InputState {
    constructor() {
        this.keyPresses = [];
        this.keyPressHandler = null;
        this.activeStream = KEYBOARD_STREAM;
    }
    init() {
        this.keyPresses = [];
        this.activeStream = KEYBOARD_STREAM;
        this.keyPressHandler = null;
    }
    keyPress(zsciiInputCode) {
        const handler = this.keyPressHandler;
        if (handler === null) {
            this.keyPresses.push(zsciiInputCode);
        }
        else {
            this.keyPressHandler = null;
            handler(zsciiInputCode);
        }
    }
    setKeyPressHandler(keyPressHandler) {
        const keyPress = this.keyPresses.shift();
        if (keyPress === undefined) {
            this.keyPressHandler = keyPressHandler;
        }
        else {
            keyPressHandler(keyPress);
        }
    }
    async readKeyPress(timer) {
        const p = new Promise((resolve) => {
            this.setKeyPressHandler(resolve);
        });
        if (timer === null) {
            return p;
        }
        const result = await Promise.race([p, timer]);
        const handler = this.keyPressHandler;
        if (handler !== null) {
            // if the handler is still set, the timer won the race
            this.keyPressHandler = null;
            handler(result);
        }
        return p;
    }
    setActiveStream(type) {
        if (type !== KEYBOARD_STREAM && type !== FILE_STREAM) {
            throw new Error(`Unexpected input stream type ${type}`);
        }
        this.activeStream = type;
    }
    getActiveStream() {
        return this.activeStream;
    }
}
const singleton = new InputState();
export default singleton;
