// zmachine-core/dist/state/memory.js
import { BYTE_MASK_0, BYTE_MASK_1 } from '../utils/constants';
// singleton for main memory storage
class Memory {
    constructor() {
        this.memory = [];
        this.length = 0;
        this.static = 0;
    }
    setMemory(file) {
        if (file.length < 64) {
            throw new Error('file has less than 64 bytes');
        }
        if (file.some((byte) => (!Number.isInteger(byte)
            || byte < 0
            || byte >= 256))) {
            throw new Error('file has non-byte values');
        }
        this.memory = file.slice();
        this.length = this.memory.length;
        this.static = this.getWord(14);
        if (this.static > this.length) {
            throw new Error('dynamic memory is larger than the file');
        }
    }
    getByte(byteAddr) {
        if (byteAddr >= this.length) {
            throw new Error('address out of range in getByte');
        }
        return this.memory[byteAddr];
    }
    setByte(byteAddr, byte) {
        if (byteAddr >= this.static) {
            throw new Error('address out of range in setByte');
        }
        if (byte < 0 || byte >= 256) {
            throw new Error('byte out of range in setByte');
        }
        this.memory[byteAddr] = byte;
    }
    getWord(byteAddr) {
        if (byteAddr + 1 >= this.length) {
            throw new Error('address out of range in getWord');
        }
        // eslint-disable-next-line no-bitwise
        return (this.memory[byteAddr] << 8) | this.memory[byteAddr + 1];
    }
    setWord(byteAddr, word) {
        if (byteAddr + 1 >= this.static) {
            throw new Error('address out of range in setWord');
        }
        if (word < 0 || word >= 65536) {
            throw new Error('word out of range in setWord');
        }
        // eslint-disable-next-line no-bitwise
        this.memory[byteAddr] = (word & BYTE_MASK_1) >> 8;
        // eslint-disable-next-line no-bitwise
        this.memory[byteAddr + 1] = (word & BYTE_MASK_0);
    }
    getBytes(byteAddr, len) {
        if (len < 0) {
            throw new Error('negative length');
        }
        if (byteAddr + len > this.length) {
            throw new Error('address out of range in getBytes');
        }
        return this.memory.slice(byteAddr, byteAddr + len);
    }
    setBytes(byteAddr, bytes) {
        const len = bytes.length;
        if (byteAddr + len > this.static) {
            throw new Error('address out of range in setBytes');
        }
        if (bytes.some((byte) => (!Number.isInteger(byte)
            || byte < 0
            || byte >= 256))) {
            throw new Error('bytes has non-byte values');
        }
        this.memory.splice(byteAddr, len, ...bytes);
    }
    getDynamicMemory() {
        return this.memory.slice(0, this.static);
    }
    setDynamicMemory(memory) {
        if (memory.length !== this.static) {
            throw new Error('Unexpected dynamic memory length');
        }
        if (memory.some((byte) => (!Number.isInteger(byte)
            || byte < 0
            || byte >= 256))) {
            throw new Error('memory has non-byte values');
        }
        this.memory.splice(0, memory.length, ...memory);
    }
}
const singleton = new Memory();
export default singleton;
