// zmachine-core/dist/state/random.js
import { BIT_15 } from '../utils/constants';
const MOD = 2147483647;
const A = 48271;
// singleton for the random number generator
class Random {
    constructor() {
        this.mode = 'RANDOM';
        this.last = 0;
        this.max = 1;
    }
    init() {
        this.mode = 'RANDOM';
        this.last = 0;
        this.max = 1;
    }
    random(range) {
        if (range < 1 || range >= BIT_15) {
            throw new Error(`range out of range ${range}`);
        }
        if (this.mode === 'RANDOM') {
            const value = Math.floor(Math.random() * range) % range;
            return value + 1;
        }
        if (this.mode === 'SEQUENTIAL') {
            this.last = (this.last % this.max) + 1;
            const value = (this.last - 1) % range;
            return value + 1;
        }
        // use MINSTD PRNG
        this.last = (this.last * A) % MOD;
        const value = (this.last) % range;
        return value + 1;
    }
    seed(value) {
        if (value < 0) {
            throw new Error('negative seed value');
        }
        if (value === 0) {
            this.mode = 'RANDOM';
        }
        else if (value < 1000) {
            this.max = value;
            this.last = value;
            this.mode = 'SEQUENTIAL';
        }
        else {
            this.last = value;
            this.mode = 'SEEDED';
        }
    }
}
const singleton = new Random();
export default singleton;
