// zmachine-core/dist/state/screen.js
const BLACK = '#000000';
const WHITE = '#ffffff';
export const NORMAL_FONT = 1;
export const GRAPHICS_FONT = 3;
export const MONOSPACE_FONT = 4;
class ScreenState {
    constructor() {
        this.screen = {
            serial: 0,
            charHeight: 16,
            charWidth: 8,
            rows: 24,
            columns: 80,
            foreground: WHITE,
            background: BLACK,
            status: {
                serial: 0,
                display: true,
                name: '',
                progress: '',
            },
            upperWindow: {
                serial: 0,
                rows: 0,
                lines: [],
            },
            lowerWindow: {
                serial: 0,
                rows: 23,
                lines: [{
                        key: 0,
                        serial: 0,
                        chars: [
                            {
                                kind: 'BLANK',
                                key: 1,
                                background: BLACK,
                                monospace: false,
                                cursor: true,
                            },
                        ],
                    }],
            },
        };
        this.upperCursor = { x: 0, y: 0 };
        this.lowerCursor = { x: 0, y: 0 };
        this.activeWindow = 'LOWER';
        this.foregroundDefault = WHITE;
        this.backgroundDefault = BLACK;
        this.foreground = WHITE;
        this.background = BLACK;
        this.font = NORMAL_FONT;
        this.upperTrueRows = 0;
        this.upperSeenRows = 0;
        this.lastKey = 1;
        this.fontStyle = {
            reverse: false,
            monospace: false,
            bold: false,
            italic: false,
        };
        this.setter = () => { };
    }
    init() {
        // resets the state for a load of a new game
        // or a restart except things which gets reset
        // via ClearScreen which preserves the screen's
        // rows and columns and knows what version of
        // game is being played
        this.foregroundDefault = WHITE;
        this.backgroundDefault = BLACK;
        this.foreground = WHITE;
        this.background = BLACK;
        this.font = NORMAL_FONT;
        this.fontStyle = {
            reverse: false,
            monospace: false,
            bold: false,
            italic: false,
        };
    }
    attachSetter(setter) {
        this.setter = setter;
    }
    detachSetter() {
        this.setter = () => { };
    }
    getScreen() {
        return this.screen;
    }
    getNextKey() {
        this.lastKey += 1;
        return this.lastKey;
    }
    flush() {
        this.setter({ ...this.screen });
    }
}
const singleton = new ScreenState();
export default singleton;
