// zmachine-core/dist/ops/dispatch.js
import GameOps from './game';
import MathOps from './math';
import MemoryOps from './memory';
import ObjectOps from './object';
import OutputOps from './output';
import PrintOps from './print';
import InputOps from './input';
import RoutineOps from './routine';
import ScreenOps from './screen';
import VariableOps from './variable';
import { LARGE_ARG, SMALL_ARG, VARIABLE_ARG, ARGS_0_RETURN, ARGS_0_VOID, ARGS_1_RETURN, ARGS_1_VOID, ARGS_2_RETURN, ARGS_2_VOID, ARGS_4_RETURN, ARGS_4_VOID, ARGS_8_VOID, } from './types';
const OPS = [];
// Two-operand opcodes 2OP
// 2OP:1 1 je a b c d ?(label)
OPS[1] = {
    kind: ARGS_2_RETURN,
    name: 'je',
    method: MathOps.je,
    types: [SMALL_ARG, SMALL_ARG],
    store: false,
    branch: true,
    version: 1,
};
// 2OP:2 2 jl a b ?(label)
OPS[2] = {
    kind: ARGS_2_RETURN,
    name: 'jl',
    method: MathOps.jl,
    types: [SMALL_ARG, SMALL_ARG],
    store: false,
    branch: true,
    version: 1,
};
// 2OP:3 3 jg a b ?(label)
OPS[3] = {
    kind: ARGS_2_RETURN,
    name: 'jg',
    method: MathOps.jg,
    types: [SMALL_ARG, SMALL_ARG],
    store: false,
    branch: true,
    version: 1,
};
// 2OP:4 4 dec_chk (variable) value ?(label)
OPS[4] = {
    kind: ARGS_2_RETURN,
    name: 'dec_chk',
    method: VariableOps.dec_chk,
    types: [SMALL_ARG, SMALL_ARG],
    store: false,
    branch: true,
    version: 1,
};
// 2OP:5 5 inc_chk (variable) value ?(label)
OPS[5] = {
    kind: ARGS_2_RETURN,
    name: 'inc_chk',
    method: VariableOps.inc_chk,
    types: [SMALL_ARG, SMALL_ARG],
    store: false,
    branch: true,
    version: 1,
};
// 2OP:6 6 jin obj1 obj2 ?(label)
OPS[6] = {
    kind: ARGS_2_RETURN,
    name: 'jin',
    method: ObjectOps.jin,
    types: [SMALL_ARG, SMALL_ARG],
    store: false,
    branch: true,
    version: 1,
};
// 2OP:7 7 test bitmap flags ?(label)
OPS[7] = {
    kind: ARGS_2_RETURN,
    name: 'test',
    method: MathOps.test,
    types: [SMALL_ARG, SMALL_ARG],
    store: false,
    branch: true,
    version: 1,
};
// 2OP:8 8 or a b -> (result)
OPS[8] = {
    kind: ARGS_2_RETURN,
    name: 'or',
    method: MathOps.or,
    types: [SMALL_ARG, SMALL_ARG],
    store: true,
    branch: false,
    version: 1,
};
// 2OP:9 9 and a b -> (result)
OPS[9] = {
    kind: ARGS_2_RETURN,
    name: 'and',
    method: MathOps.and,
    types: [SMALL_ARG, SMALL_ARG],
    store: true,
    branch: false,
    version: 1,
};
// 2OP:10 A test_attr object attribute ?(label)
OPS[10] = {
    kind: ARGS_2_RETURN,
    name: 'test_attr',
    method: ObjectOps.test_attr,
    types: [SMALL_ARG, SMALL_ARG],
    store: false,
    branch: true,
    version: 1,
};
// 2OP:11 B set_attr object attribute
OPS[11] = {
    kind: ARGS_2_VOID,
    name: 'set_attr',
    method: ObjectOps.set_attr,
    types: [SMALL_ARG, SMALL_ARG],
    store: false,
    branch: false,
    version: 1,
};
// 2OP:12 C clear_attr object attribute
OPS[12] = {
    kind: ARGS_2_VOID,
    name: 'clear_attr',
    method: ObjectOps.clear_attr,
    types: [SMALL_ARG, SMALL_ARG],
    store: false,
    branch: false,
    version: 1,
};
// 2OP:13 D store (variable) value
OPS[13] = {
    kind: ARGS_2_VOID,
    name: 'store',
    method: VariableOps.store,
    types: [SMALL_ARG, SMALL_ARG],
    store: false,
    branch: false,
    version: 1,
};
// 2OP:14 E insert_obj object destination
OPS[14] = {
    kind: ARGS_2_VOID,
    name: 'insert_obj',
    method: ObjectOps.insert_obj,
    types: [SMALL_ARG, SMALL_ARG],
    store: false,
    branch: false,
    version: 1,
};
// 2OP:15 F loadw array word-index -> (result)
OPS[15] = {
    kind: ARGS_2_RETURN,
    name: 'loadw',
    method: MemoryOps.loadw,
    types: [SMALL_ARG, SMALL_ARG],
    store: true,
    branch: false,
    version: 1,
};
// 2OP:16 10 loadb array byte-index -> (result)
OPS[16] = {
    kind: ARGS_2_RETURN,
    name: 'loadb',
    method: MemoryOps.loadb,
    types: [SMALL_ARG, SMALL_ARG],
    store: true,
    branch: false,
    version: 1,
};
// 2OP:17 11 get_prop object property -> (result)
OPS[17] = {
    kind: ARGS_2_RETURN,
    name: 'get_prop',
    method: ObjectOps.get_prop,
    types: [SMALL_ARG, SMALL_ARG],
    store: true,
    branch: false,
    version: 1,
};
// 2OP:18 12 get_prop_addr object property -> (result)
OPS[18] = {
    kind: ARGS_2_RETURN,
    name: 'get_prop_addr',
    method: ObjectOps.get_prop_addr,
    types: [SMALL_ARG, SMALL_ARG],
    store: true,
    branch: false,
    version: 1,
};
// 2OP:19 13 get_next_prop object property -> (result)
OPS[19] = {
    kind: ARGS_2_RETURN,
    name: 'get_next_prop',
    method: ObjectOps.get_next_prop,
    types: [SMALL_ARG, SMALL_ARG],
    store: true,
    branch: false,
    version: 1,
};
// 2OP:20 14 add a b -> (result)
OPS[20] = {
    kind: ARGS_2_RETURN,
    name: 'add',
    method: MathOps.add,
    types: [SMALL_ARG, SMALL_ARG],
    store: true,
    branch: false,
    version: 1,
};
// 2OP:21 15 sub a b -> (result)
OPS[21] = {
    kind: ARGS_2_RETURN,
    name: 'sub',
    method: MathOps.sub,
    types: [SMALL_ARG, SMALL_ARG],
    store: true,
    branch: false,
    version: 1,
};
// 2OP:22 16 mul a b -> (result)
OPS[22] = {
    kind: ARGS_2_RETURN,
    name: 'mul',
    method: MathOps.mul,
    types: [SMALL_ARG, SMALL_ARG],
    store: true,
    branch: false,
    version: 1,
};
// 2OP:23 17 div a b -> (result)
OPS[23] = {
    kind: ARGS_2_RETURN,
    name: 'div',
    method: MathOps.div,
    types: [SMALL_ARG, SMALL_ARG],
    store: true,
    branch: false,
    version: 1,
};
// 2OP:24 18 mod a b -> (result)
OPS[24] = {
    kind: ARGS_2_RETURN,
    name: 'mod',
    method: MathOps.mod,
    types: [SMALL_ARG, SMALL_ARG],
    store: true,
    branch: false,
    version: 1,
};
// 2OP:25 19 4 call_2s routine arg1 -> (result)
OPS[25] = {
    kind: ARGS_2_VOID,
    name: 'call_2s',
    method: RoutineOps.call_2s,
    types: [SMALL_ARG, SMALL_ARG],
    store: false,
    branch: false,
    version: 4,
};
// 2OP:26 1A 5 call_2n routine arg1
OPS[26] = {
    kind: ARGS_2_VOID,
    name: 'call_2n',
    method: RoutineOps.call_2n,
    types: [SMALL_ARG, SMALL_ARG],
    store: false,
    branch: false,
    version: 5,
};
// 2OP:27 1B 5 set_colour foreground background
OPS[27] = {
    kind: ARGS_2_VOID,
    name: 'set_colour',
    method: ScreenOps.set_colour,
    types: [SMALL_ARG, SMALL_ARG],
    store: false,
    branch: false,
    version: 5,
};
// 2OP:28 1C 5/6 throw value stack-frame
OPS[28] = {
    kind: ARGS_2_VOID,
    name: 'throw',
    method: RoutineOps.throw,
    types: [SMALL_ARG, SMALL_ARG],
    store: false,
    branch: false,
    version: 5,
};
// Opcode numbers 32 to 127: other forms of 2OP with different types
for (let opcode = 0; opcode < 32; opcode += 1) {
    if (OPS[opcode]) {
        const smallSmallOp = OPS[opcode];
        if (smallSmallOp.kind !== ARGS_2_RETURN && smallSmallOp.kind !== ARGS_2_VOID) {
            throw new Error('Unexpected kind in 2OP codes');
        }
        const smallVarOp = { ...smallSmallOp };
        const varSmallOp = { ...smallSmallOp };
        const varVarOp = { ...smallSmallOp };
        smallVarOp.types = [SMALL_ARG, VARIABLE_ARG];
        varSmallOp.types = [VARIABLE_ARG, SMALL_ARG];
        varVarOp.types = [VARIABLE_ARG, VARIABLE_ARG];
        OPS[opcode + 32] = smallVarOp;
        OPS[opcode + 64] = varSmallOp;
        OPS[opcode + 96] = varVarOp;
    }
}
// One-operand opcodes 1OP
// 1OP:128 0 jz a ?(label)
OPS[128] = {
    kind: ARGS_1_RETURN,
    name: 'jz',
    method: MathOps.jz,
    types: [LARGE_ARG],
    store: false,
    branch: true,
    version: 1,
};
// 1OP:129 1 get_sibling object -> (result) ?(label)
OPS[129] = {
    kind: ARGS_1_RETURN,
    name: 'get_sibling',
    method: ObjectOps.get_sibling,
    types: [LARGE_ARG],
    store: true,
    branch: true,
    version: 1,
};
// 1OP:130 2 get_child object -> (result) ?(label)
OPS[130] = {
    kind: ARGS_1_RETURN,
    name: 'get_child',
    method: ObjectOps.get_child,
    types: [LARGE_ARG],
    store: true,
    branch: true,
    version: 1,
};
// 1OP:131 3 get_parent object -> (result)
OPS[131] = {
    kind: ARGS_1_RETURN,
    name: 'get_parent',
    method: ObjectOps.get_parent,
    types: [LARGE_ARG],
    store: true,
    branch: false,
    version: 1,
};
// 1OP:132 4 get_prop_len property-address -> (result)
OPS[132] = {
    kind: ARGS_1_RETURN,
    name: 'get_prop_len',
    method: ObjectOps.get_prop_len,
    types: [LARGE_ARG],
    store: true,
    branch: false,
    version: 1,
};
// 1OP:133 5 inc (variable)
OPS[133] = {
    kind: ARGS_1_VOID,
    name: 'inc',
    method: VariableOps.inc,
    types: [LARGE_ARG],
    store: false,
    branch: false,
    version: 1,
};
// 1OP:134 6 dec (variable)
OPS[134] = {
    kind: ARGS_1_VOID,
    name: 'dec',
    method: VariableOps.dec,
    types: [LARGE_ARG],
    store: false,
    branch: false,
    version: 1,
};
// 1OP:135 7 print_addr byte-address-of-string
OPS[135] = {
    kind: ARGS_1_VOID,
    name: 'print_addr',
    method: PrintOps.print_addr,
    types: [LARGE_ARG],
    store: false,
    branch: false,
    version: 1,
};
// 1OP:136 8 4 call_1s routine -> (result)
OPS[136] = {
    kind: ARGS_1_VOID,
    name: 'call_1s',
    method: RoutineOps.call_1s,
    types: [LARGE_ARG],
    store: false,
    branch: false,
    version: 4,
};
// 1OP:137 9 remove_obj object
OPS[137] = {
    kind: ARGS_1_VOID,
    name: 'remove_obj',
    method: ObjectOps.remove_obj,
    types: [LARGE_ARG],
    store: false,
    branch: false,
    version: 1,
};
// 1OP:138 A print_obj object
OPS[138] = {
    kind: ARGS_1_VOID,
    name: 'print_obj',
    method: PrintOps.print_obj,
    types: [LARGE_ARG],
    store: false,
    branch: false,
    version: 1,
};
// 1OP:139 B ret value
OPS[139] = {
    kind: ARGS_1_VOID,
    name: 'ret',
    method: RoutineOps.ret,
    types: [LARGE_ARG],
    store: false,
    branch: false,
    version: 1,
};
// 1OP:140 C jump ?(label)
OPS[140] = {
    kind: ARGS_1_VOID,
    name: 'jump',
    method: RoutineOps.jump,
    types: [LARGE_ARG],
    store: false,
    branch: false,
    version: 1,
};
// 1OP:141 D print_paddr packed-address-of-string
OPS[141] = {
    kind: ARGS_1_VOID,
    name: 'print_paddr',
    method: PrintOps.print_paddr,
    types: [LARGE_ARG],
    store: false,
    branch: false,
    version: 1,
};
// 1OP:142 E load (variable) -> (result)
OPS[142] = {
    kind: ARGS_1_RETURN,
    name: 'load',
    method: VariableOps.load,
    types: [LARGE_ARG],
    store: true,
    branch: false,
    version: 1,
};
// 1OP:143 F 1/4 not value -> (result)
// 1OP:143 F 5 call_1n routine
// this opcode is handled in an unusual way because it changed in version 5
OPS[143] = {
    kind: ARGS_1_VOID,
    name: 'not/call_1n',
    method: RoutineOps.call_1n,
    types: [LARGE_ARG],
    store: false,
    branch: false,
    version: 1,
};
// Opcode numbers 144 to 175: other forms of 1OP with different types.
for (let opcode = 128; opcode < 144; opcode += 1) {
    if (OPS[opcode]) {
        const largeOp = OPS[opcode];
        if (largeOp.kind !== ARGS_1_RETURN && largeOp.kind !== ARGS_1_VOID) {
            throw new Error('Unexpected kind in 1OP codes');
        }
        const smallOp = { ...largeOp };
        const varOp = { ...largeOp };
        smallOp.types = [SMALL_ARG];
        varOp.types = [VARIABLE_ARG];
        OPS[opcode + 16] = smallOp;
        OPS[opcode + 32] = varOp;
    }
}
// Zero-operand opcodes 0OP
// 0OP:176 0 rtrue
OPS[176] = {
    kind: ARGS_0_VOID,
    name: 'rtrue',
    method: RoutineOps.rtrue,
    store: false,
    branch: false,
    version: 1,
};
// 0OP:177 1 rfalse
OPS[177] = {
    kind: ARGS_0_VOID,
    name: 'rfalse',
    method: RoutineOps.rfalse,
    store: false,
    branch: false,
    version: 1,
};
// 0OP:178 2 print <literal-string>
OPS[178] = {
    kind: ARGS_0_VOID,
    name: 'print',
    method: PrintOps.print,
    store: false,
    branch: false,
    version: 1,
};
// 0OP:179 3 print_ret <literal-string>
OPS[179] = {
    kind: ARGS_0_VOID,
    name: 'print_ret',
    method: PrintOps.print_ret,
    store: false,
    branch: false,
    version: 1,
};
// 0OP:180 4 1/- nop
OPS[180] = {
    kind: ARGS_0_VOID,
    name: 'nop',
    method: () => { },
    store: false,
    branch: false,
    version: 1,
};
// 0OP:181 5 1 save ?(label)
// 0OP:181 5 4 save -> (result)
OPS[181] = {
    kind: ARGS_0_VOID,
    name: 'save',
    method: GameOps.save,
    store: false,
    branch: false,
    version: 1,
};
// 0OP:182 6 1 restore ?(label)
// 0OP:182 5 4 restore -> (result)
OPS[182] = {
    kind: ARGS_0_VOID,
    name: 'restore',
    method: GameOps.restore,
    store: false,
    branch: false,
    version: 1,
};
// 0OP:183 7 1 restart
OPS[183] = {
    kind: ARGS_0_VOID,
    name: 'restart',
    method: GameOps.restart,
    store: false,
    branch: false,
    version: 1,
};
// 0OP:184 8 ret_popped
OPS[184] = {
    kind: ARGS_0_VOID,
    name: 'ret_popped',
    method: RoutineOps.ret_popped,
    store: false,
    branch: false,
    version: 1,
};
// 0OP:185 9 1 pop
// 0OP:185 9 5/6 catch -> (result)
// this opcode is handled in an unusual way because it changed in version 5
OPS[185] = {
    kind: ARGS_0_VOID,
    name: 'pop/catch',
    method: RoutineOps.catch,
    store: false,
    branch: false,
    version: 1,
};
// 0OP:186 A quit
OPS[186] = {
    kind: ARGS_0_VOID,
    name: 'quit',
    method: GameOps.quit,
    store: false,
    branch: false,
    version: 1,
};
// 0OP:187 B new_line
OPS[187] = {
    kind: ARGS_0_VOID,
    name: 'new_line',
    method: PrintOps.new_line,
    store: false,
    branch: false,
    version: 1,
};
// 0OP:188 C 3 show_status
OPS[188] = {
    kind: ARGS_0_VOID,
    name: 'show_status',
    method: PrintOps.show_status,
    store: false,
    branch: false,
    version: 3,
};
// 0OP:189 D 3 verify ?(label)
OPS[189] = {
    kind: ARGS_0_RETURN,
    name: 'verify',
    method: GameOps.verify,
    store: false,
    branch: true,
    version: 3,
};
// 0OP:191 F 5/- piracy ?(label)
OPS[191] = {
    kind: ARGS_0_RETURN,
    name: 'piracy',
    method: GameOps.piracy,
    store: false,
    branch: true,
    version: 5,
};
// Opcode numbers 192 to 223: VAR forms of 2OP:0 to 2OP:31.
for (let opcode = 0; opcode < 32; opcode += 1) {
    if (OPS[opcode]) {
        const longFormOp = OPS[opcode];
        if (longFormOp.kind === ARGS_2_RETURN) {
            const varFormOp = {
                kind: ARGS_4_RETURN,
                name: longFormOp.name,
                method: longFormOp.method,
                store: longFormOp.store,
                branch: longFormOp.branch,
                version: longFormOp.version,
            };
            OPS[opcode + 192] = varFormOp;
        }
        else if (longFormOp.kind === ARGS_2_VOID) {
            const varFormOp = {
                kind: ARGS_4_VOID,
                name: longFormOp.name,
                method: longFormOp.method,
                store: longFormOp.store,
                branch: longFormOp.branch,
                version: longFormOp.version,
            };
            OPS[opcode + 192] = varFormOp;
        }
        else {
            throw new Error('Unexpected kind in 2OP codes');
        }
    }
}
// Variable-operand opcodes VAR
// VAR:224 0 1 call routine ...0 to 3 args... -> (result)
OPS[224] = {
    kind: ARGS_4_VOID,
    name: 'call_vs',
    method: RoutineOps.call_vs,
    store: false,
    branch: false,
    version: 1,
};
// VAR:225 1 storew array word-index value
OPS[225] = {
    kind: ARGS_4_VOID,
    name: 'storew',
    method: MemoryOps.storew,
    store: false,
    branch: false,
    version: 1,
};
// VAR:226 2 storeb array byte-index value
OPS[226] = {
    kind: ARGS_4_VOID,
    name: 'storeb',
    method: MemoryOps.storeb,
    store: false,
    branch: false,
    version: 1,
};
// VAR:227 3 put_prop object property value
OPS[227] = {
    kind: ARGS_4_VOID,
    name: 'put_prop',
    method: ObjectOps.put_prop,
    store: false,
    branch: false,
    version: 1,
};
// VAR:228 4 1 sread text parse time routine
OPS[228] = {
    kind: ARGS_4_VOID,
    name: 'read',
    method: InputOps.read,
    store: false,
    branch: false,
    version: 1,
};
// VAR:229 5 print_char output-character-code
OPS[229] = {
    kind: ARGS_4_VOID,
    name: 'print_char',
    method: PrintOps.print_char,
    store: false,
    branch: false,
    version: 1,
};
// VAR:230 6 print_num value
OPS[230] = {
    kind: ARGS_4_VOID,
    name: 'print_num',
    method: PrintOps.print_num,
    store: false,
    branch: false,
    version: 1,
};
// VAR:231 7 random range -> (result)
OPS[231] = {
    kind: ARGS_4_RETURN,
    name: 'random',
    method: MathOps.random,
    store: true,
    branch: false,
    version: 1,
};
// VAR:232 8 push value
OPS[232] = {
    kind: ARGS_4_VOID,
    name: 'push',
    method: VariableOps.push,
    store: false,
    branch: false,
    version: 1,
};
// VAR:233 9 1 pull (variable)
OPS[233] = {
    kind: ARGS_4_VOID,
    name: 'pull',
    method: VariableOps.pull,
    store: false,
    branch: false,
    version: 1,
};
// VAR:234 A 3 split_window lines
OPS[234] = {
    kind: ARGS_4_VOID,
    name: 'split_window',
    method: ScreenOps.split_window,
    store: false,
    branch: false,
    version: 3,
};
// VAR:235 B 3 set_window window
OPS[235] = {
    kind: ARGS_4_VOID,
    name: 'set_window',
    method: ScreenOps.set_window,
    store: false,
    branch: false,
    version: 3,
};
// VAR:236 C 4 call_vs2 routine ...up to 7 args... -> (result)
OPS[236] = {
    kind: ARGS_8_VOID,
    name: 'call_vs2',
    method: RoutineOps.call_vs2,
    store: false,
    branch: false,
    version: 4,
};
// VAR:237 D 4 erase_window window
OPS[237] = {
    kind: ARGS_4_VOID,
    name: 'erase_window',
    method: ScreenOps.erase_window,
    store: false,
    branch: false,
    version: 4,
};
// VAR:238 E 4/6 erase_line value
OPS[238] = {
    kind: ARGS_4_VOID,
    name: 'erase_line',
    method: ScreenOps.erase_line,
    store: false,
    branch: false,
    version: 4,
};
// VAR:239 F 4 set_cursor line column
OPS[239] = {
    kind: ARGS_4_VOID,
    name: 'set_cursor',
    method: ScreenOps.set_cursor,
    store: false,
    branch: false,
    version: 4,
};
// VAR:240 10 4/6 get_cursor array
OPS[240] = {
    kind: ARGS_4_VOID,
    name: 'get_cursor',
    method: ScreenOps.get_cursor,
    store: false,
    branch: false,
    version: 4,
};
// VAR:241 11 4 set_text_style style
OPS[241] = {
    kind: ARGS_4_VOID,
    name: 'set_text_style',
    method: ScreenOps.set_text_style,
    store: false,
    branch: false,
    version: 4,
};
// VAR:242 12 4 buffer_mode flag
OPS[242] = {
    kind: ARGS_4_VOID,
    name: 'buffer_mode',
    method: OutputOps.buffer_mode,
    store: false,
    branch: false,
    version: 4,
};
// VAR:243 13 3 output_stream number table
OPS[243] = {
    kind: ARGS_4_VOID,
    name: 'output_stream',
    method: OutputOps.output_stream,
    store: false,
    branch: false,
    version: 3,
};
// VAR:244 14 3 input_stream number
OPS[244] = {
    kind: ARGS_4_VOID,
    name: 'input_stream',
    method: InputOps.input_stream,
    store: false,
    branch: false,
    version: 3,
};
// VAR:245 15 5/3 sound_effect number effect volume routine
OPS[245] = {
    kind: ARGS_4_VOID,
    name: 'sound_effect',
    method: OutputOps.sound_effect,
    store: false,
    branch: false,
    version: 3,
};
// VAR:246 16 4 read_char 1 time routine -> (result)
OPS[246] = {
    kind: ARGS_4_VOID,
    name: 'read_char',
    method: InputOps.read_char,
    store: false,
    branch: false,
    version: 4,
};
// VAR:247 17 4 scan_table x table len form -> (result)
OPS[247] = {
    kind: ARGS_4_RETURN,
    name: 'scan_table',
    method: MemoryOps.scan_table,
    store: true,
    branch: true,
    version: 4,
};
// VAR:248 18 5/6 not value -> (result)
OPS[248] = {
    kind: ARGS_4_RETURN,
    name: 'not',
    method: MathOps.not,
    store: true,
    branch: false,
    version: 5,
};
// VAR:249 19 5 call_vn routine ...up to 3 args...
OPS[249] = {
    kind: ARGS_4_VOID,
    name: 'call_vn',
    method: RoutineOps.call_vn,
    store: false,
    branch: false,
    version: 5,
};
// VAR:250 1A 5 call_vn2 routine ...up to 7 args...
OPS[250] = {
    kind: ARGS_8_VOID,
    name: 'call_vn2',
    method: RoutineOps.call_vn2,
    store: false,
    branch: false,
    version: 5,
};
// VAR:251 1B 5 tokenise text parse dictionary flag
OPS[251] = {
    kind: ARGS_4_VOID,
    name: 'tokenise',
    method: InputOps.tokenise,
    store: false,
    branch: false,
    version: 5,
};
// VAR:252 1C 5 encode_text zscii-text length from coded-text
OPS[252] = {
    kind: ARGS_4_VOID,
    name: 'encode_text',
    method: InputOps.encode_text,
    store: false,
    branch: false,
    version: 5,
};
// VAR:253 1D 5 copy_table first second size
OPS[253] = {
    kind: ARGS_4_VOID,
    name: 'copy_table',
    method: MemoryOps.copy_table,
    store: false,
    branch: false,
    version: 5,
};
// VAR:254 1E 5 print_table zscii-text width height skip
OPS[254] = {
    kind: ARGS_4_VOID,
    name: 'print_table',
    method: PrintOps.print_table,
    store: false,
    branch: false,
    version: 5,
};
// VAR:255 1F 5 check_arg_count argument-number
OPS[255] = {
    kind: ARGS_4_RETURN,
    name: 'check_arg_count',
    method: RoutineOps.check_arg_count,
    store: false,
    branch: true,
    version: 5,
};
// Extended opcodes EXT
// EXT:0 0 5 save table bytes name prompt -> (result)
OPS[256] = {
    kind: ARGS_4_VOID,
    name: 'save',
    method: GameOps.save_v5,
    store: false,
    branch: false,
    version: 5,
};
// EXT:1 1 5 restore table bytes name prompt-> (result)
OPS[257] = {
    kind: ARGS_4_VOID,
    name: 'restore',
    method: GameOps.restore_v5,
    store: false,
    branch: false,
    version: 5,
};
// EXT:2 2 5 log_shift number places -> (result)
OPS[258] = {
    kind: ARGS_4_RETURN,
    name: 'log_shift',
    method: MathOps.log_shift,
    store: true,
    branch: false,
    version: 5,
};
// EXT:3 3 5/- art_shift number places -> (result)
OPS[259] = {
    kind: ARGS_4_RETURN,
    name: 'art_shift',
    method: MathOps.art_shift,
    store: true,
    branch: false,
    version: 5,
};
// EXT:4 4 5 set_font font -> (result)
OPS[260] = {
    kind: ARGS_4_RETURN,
    name: 'set_font',
    method: ScreenOps.set_font,
    store: true,
    branch: false,
    version: 5,
};
// EXT:9 9 5 save_undo -> (result)
OPS[265] = {
    kind: ARGS_4_RETURN,
    name: 'save_undo',
    method: GameOps.save_undo,
    store: true,
    branch: false,
    version: 5,
};
// EXT:10 A 5 restore_undo -> (result)
OPS[266] = {
    kind: ARGS_4_RETURN,
    name: 'restore_undo',
    method: GameOps.restore_undo,
    store: true,
    branch: false,
    version: 5,
};
// EXT:11 B 5/* print_unicode char-number
OPS[267] = {
    kind: ARGS_4_VOID,
    name: 'print_unicode',
    method: PrintOps.print_unicode,
    store: false,
    branch: false,
    version: 5,
};
// EXT:12 C 5/* check_unicode char-number -> (result)
OPS[268] = {
    kind: ARGS_4_RETURN,
    name: 'check_unicode',
    method: PrintOps.check_unicode,
    store: true,
    branch: false,
    version: 5,
};
// EXT:13 D 5/* set_true_colour foreground background
OPS[269] = {
    kind: ARGS_4_VOID,
    name: 'set_true_colour',
    method: ScreenOps.set_true_colour,
    store: false,
    branch: false,
    version: 5,
};
export default OPS;
