// zmachine-core/dist/ops/game.js
// OP codes which deal with the game state are defined here
import Game from '../resources/game';
import Header from '../resources/header';
import Routine from '../resources/routine';
import ScreenState from '../state/screen';
import { isBadArg, throwBadArg, QUIT, } from './types';
/* eslint-disable @typescript-eslint/camelcase */
export default class GameOps {
    static async save() {
        const version = Header.version();
        const success = await Game.saveGame();
        if (version >= 4) {
            if (success) {
                Routine.store(1);
            }
            else {
                Routine.store(0);
            }
        }
        else if (success) {
            await Routine.branch(1);
        }
        else {
            await Routine.branch(0);
        }
    }
    static async save_v5(table, bytes, name, prompt) {
        let success = false;
        if (table === null) {
            success = await Game.saveGame();
        }
        else {
            if (isBadArg(bytes) || isBadArg(name)) {
                throwBadArg();
            }
            if (prompt !== 0 && prompt !== 1 && prompt !== null) {
                throw new Error(`Unexpected prompt ${prompt}`);
            }
            const doPrompt = prompt === 1;
            const doNotPrompt = prompt === 0;
            success = await Game.saveTable(table, bytes, name, doPrompt, doNotPrompt);
        }
        if (success) {
            Routine.store(1);
        }
        else {
            Routine.store(0);
        }
    }
    static async restore() {
        const version = Header.version();
        const success = await Game.restoreGame();
        if (version >= 4) {
            if (success) {
                Routine.store(2);
            }
            else {
                Routine.store(0);
            }
        }
        else if (success) {
            await Routine.branch(1);
        }
        else {
            await Routine.branch(0);
        }
    }
    static async restore_v5(table, bytes, name, prompt) {
        let storeValue = 0;
        if (table === null) {
            const success = await Game.restoreGame();
            if (success) {
                storeValue = 2;
            }
        }
        else {
            if (isBadArg(bytes) || isBadArg(name)) {
                throwBadArg();
            }
            if (prompt !== 0 && prompt !== 1 && prompt !== null) {
                throw new Error(`Unexpected prompt ${prompt}`);
            }
            const doPrompt = prompt === 1;
            const doNotPrompt = prompt === 0;
            storeValue = await Game.restoreTable(table, bytes, name, doPrompt, doNotPrompt);
        }
        Routine.store(storeValue);
    }
    static save_undo() {
        const success = Game.saveUndo();
        if (success) {
            return 1;
        }
        return 0;
    }
    static restore_undo() {
        const success = Game.restoreUndo();
        if (success) {
            return 2;
        }
        return 0;
    }
    static restart() {
        Game.restartGame();
    }
    static quit() {
        ScreenState.flush();
        ScreenState.detachSetter();
        throw new Error(QUIT);
    }
    static verify() {
        if (Game.verifyChecksum()) {
            return 1;
        }
        return 0;
    }
    static piracy() {
        return 1;
    }
}
