// zmachine-core/dist/ops/object.js
// OP codes which deal with objects are defined here
import Objects from '../resources/object';
import { isBadArg, throwBadArg } from './types';
/* eslint-disable @typescript-eslint/camelcase */
export default class ObjectOps {
    static clear_attr(object, attribute) {
        if (isBadArg(object) || isBadArg(attribute)) {
            throwBadArg();
        }
        Objects.clearAttr(object, attribute);
    }
    static get_child(object) {
        if (isBadArg(object)) {
            throwBadArg();
        }
        return Objects.getChild(object);
    }
    static get_parent(object) {
        if (isBadArg(object)) {
            throwBadArg();
        }
        return Objects.getParent(object);
    }
    static get_prop(object, property) {
        if (isBadArg(object) || isBadArg(property)) {
            throwBadArg();
        }
        return Objects.getProp(object, property);
    }
    static get_prop_addr(object, property) {
        if (isBadArg(object) || isBadArg(property)) {
            throwBadArg();
        }
        return Objects.getPropertyAddr(object, property);
    }
    static get_next_prop(object, property) {
        if (isBadArg(object) || isBadArg(property)) {
            throwBadArg();
        }
        return Objects.getNextProperty(object, property);
    }
    static get_prop_len(propertyAddress) {
        if (isBadArg(propertyAddress)) {
            throwBadArg();
        }
        return Objects.getPropertyLength(propertyAddress);
    }
    static get_sibling(object) {
        if (isBadArg(object)) {
            throwBadArg();
        }
        return Objects.getSibling(object);
    }
    static insert_obj(object, destination) {
        if (isBadArg(object) || isBadArg(destination)) {
            throwBadArg();
        }
        Objects.insertObject(object, destination);
    }
    static jin(obj1, obj2) {
        if (isBadArg(obj1) || isBadArg(obj2)) {
            throwBadArg();
        }
        const parent = Objects.getParent(obj1);
        if (obj2 === parent) {
            return 1;
        }
        return 0;
    }
    static put_prop(object, property, value) {
        if (isBadArg(object) || isBadArg(property) || isBadArg(value)) {
            throwBadArg();
        }
        Objects.putProp(object, property, value);
    }
    static remove_obj(object) {
        if (isBadArg(object)) {
            throwBadArg();
        }
        Objects.removeObject(object);
    }
    static set_attr(object, attribute) {
        if (isBadArg(object) || isBadArg(attribute)) {
            throwBadArg();
        }
        Objects.setAttr(object, attribute);
    }
    static test_attr(object, attribute) {
        if (isBadArg(object) || isBadArg(attribute)) {
            throwBadArg();
        }
        return Objects.testAttr(object, attribute);
    }
}
