// zmachine-core/dist/ops/screen.js
// OP codes which deal with the screen, windows and cursors are defined here
import Screen from '../resources/screen';
import Memory from '../state/memory';
import { isBadArg, throwBadArg } from './types';
import { BIT_15, BIT_16 } from '../utils/constants';
const COLOR_MAP = [
    -2,
    -1,
    0,
    29,
    832,
    957,
    22944,
    31775,
    30624,
    32767,
    23254,
    17969,
    11627,
];
/* eslint-disable @typescript-eslint/camelcase */
export default class ScreenOps {
    static erase_window(window) {
        if (isBadArg(window)) {
            throwBadArg();
        }
        let signedWindow = window;
        if (signedWindow >= BIT_15) {
            signedWindow -= BIT_16;
        }
        if (signedWindow === -1) {
            // clear all windows and unsplit screen
            Screen.clearScreen();
        }
        else if (signedWindow === 1) {
            Screen.clearUpperWindow();
        }
        else if (signedWindow === 0) {
            Screen.clearLowerWindow();
        }
        else if (signedWindow === -2) {
            Screen.clearUpperWindow();
            Screen.clearLowerWindow();
        }
        else {
            throw new Error(`Unexpected window erase_window(${signedWindow})`);
        }
    }
    static split_window(lines) {
        if (isBadArg(lines)) {
            throwBadArg();
        }
        Screen.splitWindow(lines);
    }
    static set_window(window) {
        if (isBadArg(window)) {
            throwBadArg();
        }
        Screen.setWindow(window);
    }
    static set_text_style(style) {
        if (isBadArg(style)) {
            throwBadArg();
        }
        Screen.setTextStyle(style);
    }
    static set_colour(foreground, background) {
        if (isBadArg(foreground) || isBadArg(background)) {
            throwBadArg();
        }
        if (foreground > 9 || background > 9) {
            throw new Error(`Unexpected colors ${foreground}, ${background}`);
        }
        Screen.setTextColor(COLOR_MAP[foreground], COLOR_MAP[background]);
    }
    static set_true_colour(foreground, background) {
        if (isBadArg(foreground) || isBadArg(background)) {
            throwBadArg();
        }
        let sForeground = foreground;
        if (sForeground >= BIT_15) {
            sForeground -= BIT_16;
        }
        let sBackground = background;
        if (sBackground >= BIT_15) {
            sBackground -= BIT_16;
        }
        Screen.setTextColor(sForeground, sBackground);
    }
    static set_font(font) {
        if (isBadArg(font)) {
            throwBadArg();
        }
        const currentFont = Screen.getFont();
        if (font === 0) {
            return currentFont;
        }
        if (font === 1 || font === 3 || font === 4) {
            Screen.setFont(font);
            return currentFont;
        }
        return 0;
    }
    static set_cursor(line, column) {
        if (isBadArg(line) || isBadArg(column)) {
            throwBadArg();
        }
        let signedLine = line;
        if (signedLine >= BIT_15) {
            signedLine -= BIT_16;
        }
        let signedColumn = column;
        if (signedColumn >= BIT_15) {
            signedColumn -= BIT_16;
        }
        Screen.setCursor(signedLine - 1, signedColumn - 1);
    }
    static get_cursor(array) {
        if (isBadArg(array)) {
            throwBadArg();
        }
        const { x, y } = Screen.getCursor();
        Memory.setWord(array, y + 1);
        Memory.setWord(array, x + 1);
    }
    static erase_line(value) {
        if (isBadArg(value)) {
            throwBadArg();
        }
        if (value === 1) {
            Screen.eraseLine();
        }
    }
}
