// zmachine-core/dist/ops/variable.js
// OP codes which deal with memory access are defined here
import Variable from '../resources/variable';
import { isBadArg, throwBadArg } from './types';
import { BIT_15, BIT_16 } from '../utils/constants';
/* eslint-disable @typescript-eslint/camelcase */
export default class VariableOps {
    static dec(variable) {
        if (isBadArg(variable)) {
            throwBadArg();
        }
        let newValue = Variable.readVar(variable) - 1;
        if (newValue < 0) {
            newValue += BIT_16;
        }
        Variable.writeVar(variable, newValue);
    }
    static dec_chk(variable, value) {
        if (isBadArg(variable) || isBadArg(value)) {
            throwBadArg();
        }
        let newValue = Variable.readVar(variable) - 1;
        if (newValue < 0) {
            newValue += BIT_16;
        }
        Variable.writeVar(variable, newValue);
        // signed comparison
        if (newValue >= BIT_15) {
            newValue -= BIT_16;
        }
        let testValue = value;
        if (testValue >= BIT_15) {
            testValue -= BIT_16;
        }
        if (newValue < testValue) {
            return 1;
        }
        return 0;
    }
    static inc(variable) {
        if (isBadArg(variable)) {
            throwBadArg();
        }
        let newValue = Variable.readVar(variable) + 1;
        if (newValue >= BIT_16) {
            newValue -= BIT_16;
        }
        Variable.writeVar(variable, newValue);
    }
    static inc_chk(variable, value) {
        if (isBadArg(variable) || isBadArg(value)) {
            throwBadArg();
        }
        let newValue = Variable.readVar(variable) + 1;
        if (newValue >= BIT_16) {
            newValue -= BIT_16;
        }
        Variable.writeVar(variable, newValue);
        // signed comparison
        if (newValue >= BIT_15) {
            newValue -= BIT_16;
        }
        let testValue = value;
        if (testValue >= BIT_15) {
            testValue -= BIT_16;
        }
        if (newValue > testValue) {
            return 1;
        }
        return 0;
    }
    static load(variable) {
        if (isBadArg(variable)) {
            throwBadArg();
        }
        const value = Variable.readVar(variable);
        if (variable === 0) {
            // This ensures that the stack is read in-place
            Variable.writeVar(variable, value);
        }
        return value;
    }
    static pull(variable) {
        if (isBadArg(variable)) {
            throwBadArg();
        }
        const value = Variable.readVar(0);
        if (variable === 0) {
            // This ensures that the stack is written in-place
            Variable.readVar(variable);
        }
        Variable.writeVar(variable, value);
    }
    static push(value) {
        if (isBadArg(value)) {
            throwBadArg();
        }
        Variable.writeVar(0, value);
    }
    static store(variable, value) {
        if (isBadArg(variable) || isBadArg(value)) {
            throwBadArg();
        }
        if (variable === 0) {
            // This ensures that the stack is written in-place
            Variable.readVar(variable);
        }
        Variable.writeVar(variable, value);
    }
}
