// ebozz/dist/Screen.js
export var TextStyle;
(function (TextStyle) {
    TextStyle[TextStyle["Roman"] = 0] = "Roman";
    TextStyle[TextStyle["ReverseVideo"] = 1] = "ReverseVideo";
    TextStyle[TextStyle["Bold"] = 2] = "Bold";
    TextStyle[TextStyle["Italic"] = 4] = "Italic";
    TextStyle[TextStyle["FixedPitch"] = 8] = "FixedPitch";
})(TextStyle = TextStyle || (TextStyle = {}));
export var Color;
(function (Color) {
    Color[Color["Current"] = 0] = "Current";
    Color[Color["Default"] = 1] = "Default";
    Color[Color["Black"] = 2] = "Black";
    Color[Color["Red"] = 3] = "Red";
    Color[Color["Green"] = 4] = "Green";
    Color[Color["Yellow"] = 5] = "Yellow";
    Color[Color["Blue"] = 6] = "Blue";
    Color[Color["Magenta"] = 7] = "Magenta";
    Color[Color["Cyan"] = 8] = "Cyan";
    Color[Color["White"] = 9] = "White";
    Color[Color["Gray"] = 10] = "Gray";
})(Color = Color || (Color = {}));
export function colorToString(c) {
    switch (c) {
        case Color.Black:
            return "black";
        case Color.Red:
            return "red";
        case Color.Green:
            return "green";
        case Color.Yellow:
            return "yellow";
        case Color.Blue:
            return "blue";
        case Color.Magenta:
            return "magenta";
        case Color.Cyan:
            return "cyan";
        case Color.White:
            return "white";
        case Color.Gray:
            return "gray";
        default:
            return "";
    }
}
export var BufferMode;
(function (BufferMode) {
    BufferMode[BufferMode["NotBuffered"] = 0] = "NotBuffered";
    BufferMode[BufferMode["Buffered"] = 1] = "Buffered";
})(BufferMode = BufferMode || (BufferMode = {}));
export class ScreenBase {
    constructor(log, id) {
        this.log = log;
        this.id = id;
    }
    getInputFromUser(_game, _input_state) {
        this.log.debug(`not implemented: ${this.id} getInputFromUser`);
    }
    getKeyFromUser(_game, _input_state) {
        this.log.debug(`not implemented: ${this.id} getInputFromUser`);
    }
    print(_game, _str) {
        this.log.debug(`not implemented: ${this.id} print`);
    }
    splitWindow(_game, lines) {
        this.log.debug(`not implemented: ${this.id} splitWindow lines=${lines}`);
    }
    setOutputWindow(_game, windowId) {
        this.log.debug(`not implemented: ${this.id} setOutputWindow windowId=${windowId}`);
    }
    getOutputWindow(_game) {
        this.log.debug(`not implemented: ${this.id} getOutputWindow`);
        return 0;
    }
    clearWindow(_game, windowId) {
        this.log.debug(`not implemented: ${this.id} clearWindow windowId=${windowId}`);
    }
    clearLine(_game, value) {
        this.log.debug(`not implemented: ${this.id} clearLine value=${value}`);
    }
    setCursorPosition(_game, line, column, windowId) {
        this.log.debug(`not implemented: ${this.id} setCursorPosition line=${line} column=${column} windowId=${windowId}`);
    }
    hideCursor(_game, windowId) {
        this.log.debug(`not implemented: ${this.id} hideCursor windowId=${windowId}`);
    }
    showCursor(_game, windowId) {
        this.log.debug(`not implemented: ${this.id} showCursor windowId=${windowId}`);
    }
    setBufferMode(_game, mode) {
        this.log.debug(`not implemented: ${this.id} setBufferMode mode=${mode}`);
    }
    setTextStyle(_game, style) {
        this.log.debug(`not implemented: ${this.id} setTextStyle style=${style}`);
    }
    setTextColors(game, windowId, foreground, background) {
        this.log.debug(`not implemented: ${this.id} setTextColors windowId=${windowId} foreground=${foreground} background=${background}`);
    }
    enableOutputStream(_game, streamId, table, width) {
        this.log.error(`not implemented: ${this.id} enableOutputStream streamId=${streamId} table=${table} width=${width}`);
    }
    disableOutputStream(_game, streamId, table, width) {
        this.log.error(`not implemented: ${this.id} disableOutputStream streamId=${streamId} table=${table} width=${width}`);
    }
    selectInputStream(_game, streamId) {
        this.log.error(`not implemented: ${this.id} selectInputStream streamId=${streamId}`);
    }
    getSize() {
        this.log.debug(`not implemented: ${this.id} getSize`);
        return { rows: 25, cols: 80 };
    }
    updateStatusBar(lhs, rhs) {
        this.log.debug(`not implemented: ${this.id} updateStatusBar lhs=${lhs} rhs=${rhs}`);
    }
    getCapabilities() {
        this.log.debug(`not implemented: ${this.id} getCapabilities`);
        return {
            hasColors: false,
            hasBold: false,
            hasItalic: false,
            hasReverseVideo: false,
            hasFixedPitch: false,
            hasSplitWindow: false,
            hasDisplayStatusBar: false,
            hasPictures: false,
            hasSound: false,
            hasTimedKeyboardInput: false,
        };
    }
}
